### DISCUSSION TITLE
Blazor App HttpClient can't fetch due to being blocked by CORS

### DISCUSSION THREAD

**Discussion by sassy224** [dotnet/aspnetcore#22988] [General] [answered]
### Describe the bug
Use HttpClient to post request to another site will result in exception
`Access to fetch at 'https://login.microsoftonline.com/xxxxxxxxxxx/oauth2/token' from origin 'https://localhost:44384' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
`
### To Reproduce
- Create a new Blazor WebAssembly App targets .Net Core 3.1 (all package versions are 3.2.0)
-Edit method IncrementCount in Counter.razor:
```
private async void IncrementCount()
    {
        currentCount++;

        HttpClient client = new HttpClient();
        var grantString = $@"grant_type=password&username=username&password=password&client_id=clientID&client_secret=clientSecret";

        grantString += $@"&resource=https://xxxx.dynamics.com/";

        var authorizationStr = Convert.ToBase64String(Encoding.ASCII.GetBytes($"clientID:clientSecret"));
        var authenticationHeader = new AuthenticationHeaderValue("Basic", authorizationStr);

        var requestMessage = new HttpRequestMessage(HttpMethod.Post, "https://login.microsoftonline.com/xxxxxxxxxxx/oauth2/token");
        if (authenticationHeader != null)
        {
            requestMessage.Headers.Authorization = authenticationHeader;
        }
        requestMessage.Content = new StringContent(grantString, Encoding.UTF8, "application/x-www-form-urlencoded");

        var response = await client.SendAsync(requestMessage).ConfigureAwait(false);
        var result1 = await response.Content.ReadAsStringAsync().ConfigureAwait(false);

    }
```
- Run blazor, navigate to Counter, press button, error is shown in console.

### Exceptions (if any)
```
Access to fetch at 'https://login.microsoftonline.com/xxxxxxxxxxx/oauth2/token' from origin 'https://localhost:44356' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```

### Further technical details
- ASP.NET Core version: 3.1
- Include the output of `dotnet --info`
```
.NET SDK (reflecting any global.json):
 Version:   5.0.100-preview.5.20279.10
 Commit:    8139f1b74e

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.100-preview.5.20279.10\

Host (useful for support):
  Version: 5.0.0-preview.5.20278.1
  Commit:  4ae4e2fe08

.NET SDKs installed:
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.500 [C:\Program Files\dotnet\sdk]
  2.1.507 [C:\Program Files\dotnet\sdk]
  2.1.508 [C:\Program Files\dotnet\sdk]
  2.1.509 [C:\Program Files\dotnet\sdk]
  2.1.511 [C:\Program Files\dotnet\sdk]
  2.1.512 [C:\Program Files\dotnet\sdk]
  2.1.513 [C:\Program Files\dotnet\sdk]
  2.1.514 [C:\Program Files\dotnet\sdk]
  2.1.515 [C:\Program Files\dotnet\sdk]
  2.1.602 [C:\Program Files\dotnet\sdk]
  2.1.700 [C:\Program Files\dotnet\sdk]
  2.1.701 [C:\Program Files\dotnet\sdk]
  2.1.801 [C:\Program Files\dotnet\sdk]
  2.1.802 [C:\Program Files\dotnet\sdk]
  2.2.207 [C:\Program Files\dotnet\sdk]
  3.1.200-preview-015002 [C:\Program Files\dotnet\sdk]
  3.1.301 [C:\Program Files\dotnet\sdk]
  3.1.400-preview-015151 [C:\Program Files\dotnet\sdk]
  5.0.100-preview.5.20279.10 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.18 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.19 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.18 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.19 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.0-preview.5.20279.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.18 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.19 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.0-preview.5.20278.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.0-preview.5.20278.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```
- The IDE (VS / VS Code/ VS4Mac) you're running on, and it's version
```
VS 2019 16.6.2
```

**Comment by davidfowl** [comment]
Is there a question? Blazor wasm has the same restrictions running JavaScript in the browser (it all runs in the same sandbox)

**Comment by sassy224** [comment]
> Is there a question? Blazor wasm has the same restrictions running JavaScript in the browser (it all runs in the same sandbox)

Are you saying that this issue is by design? I can see the same snippet works fine in Blazor server app, console app, desktop app. Can you elaborate more on why Blazor wasm doesn't work?

**Reply by javiercn** [reply]
You can enable cross-origin requests provided that the server supports CORS using the following extension method `httpRequest.SetBrowserRequestMode(BrowserRequestMode.Cors))`

**Reply by dwbidan** [reply]
This looks great! So where do I place this snippet?

**Comment by Andras-Csanyi** [comment]
> Are you saying that this issue is by design? I can see the same snippet works fine in Blazor server app, console app, desktop app. Can you elaborate more on why Blazor wasm doesn't work?

Yup. Blazor WebAssembly runs in your browser where particular security features are forced. One of them you met. If you run Blazor Server Side then your code runs (communicates other sites) at your server - which is not your browser - and send the DOM changes to your browser via SignalR. That is the difference. It is a [SO](https://stackoverflow.com/questions/10636611/how-does-access-control-allow-origin-header-work) article explains some details, there are might be better articles about it.

**Comment by sassy224** [comment]
> Yup. Blazor WebAssembly runs in your browser where particular security features are forced. One of them you met. If you run Blazor Server Side then your code runs (communicates other sites) at your server - which is not your browser - and send the DOM changes to your browser via SignalR. That is the difference. It is a [SO](https://stackoverflow.com/questions/10636611/how-does-access-control-allow-origin-header-work) article explains some details, there are might be better articles about it.

Thanks, very useful and clear for me. So in my issue, the only way it can work is the other site (https://login.microsoftonline.com/xxxxxxxxxxx/oauth2/token) must return response with header including Access-Control-Allow-Origin = "*", right?

**Comment by javiercn** [comment]
You can use CORS in Blazor Webassembly as well as configuring other `fetch` options by calling extension methods on the request. For example `request.SetBrowserRequestMode(BrowserRequestMode.Cors)` provided that the server supports CORS.

**Reply by eXapppk** [reply]
Where i can add this line into my code?

**Reply by yogyogi** [reply]
You can use CORS in Blazor Webassembly as well as configuring other fetch options ?? How to do it.

**Comment by ricksparx** [comment, marked as answer]
Hi @javiercn ,below is my code where i am setting the default browser as CORS but still getting the same error.
Could you please give us your working code repo. 
builder.Services.AddTransient(sp => new HttpClient(new DefaultBrowserOptionsMessageHandler(new WebAssemblyHttpHandler())
            {
                //DefaultBrowserRequestCache = BrowserRequestCache.NoStore,
                //DefaultBrowserRequestCredentials = BrowserRequestCredentials.Omit,
                DefaultBrowserRequestMode = BrowserRequestMode.Cors,
            })
            {
                BaseAddress = new Uri(builder.HostEnvironment.BaseAddress),
            });

**Reply by javiercn** [reply]
Is the server returning the appropriate CORS headers in response?

**Reply by ricksparx** [reply]
Thank You @javiercn I did this via proxy. I created a Proxy API and routed my endpoint via this proxy API

**Reply by eXapppk** [reply]
can you please share some content how to create Proxy API.

**Comment by michalss** [comment]
Need this as well pls

**Comment by paulomorgado** [comment]
`httpRequest.SetBrowserRequestMode(BrowserRequestMode.Cors)` is not working for me.

I'm trying to get data from an IoT device and the only thing that doesn't throw is `httpRequest.SetBrowserRequestMode(BrowserRequestMode.NoCors)`, but I always get an empty string.

The request payload is something like:

```
GET /rpc/EM1.GetStatus?id=0 HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate
Accept-Language: en-US,en;q=0.9
Connection: keep-alive
Host: 192.168.20.33
Referer: http://localhost:5229/
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36 Edg/120.0.0.0
```

And the device replies with:

```
HTTP/1.1 200 OK
Content-Type: application/json
Content-Length: 125
Server: ShellyHTTP/1.0.0
Connection: close
```

However, there's no body.

The device replies with 200 and if I issue the same request in Fiddler, I get the response body with the same 125 bytes length.

If I set Fiddler as a proxy, I can see in Fiddler the content.

**Reply by javiercn** [reply]
@paulomorgado If you are querying a different origin and they don't send back CORS requests (seems the device isn't) there's nothing you can do. Also that thing is replying with HTTP/1.0 (not even 1/1) so it's incredibly old and likely to run into all sort of weird compat issues.

**Reply by paulomorgado** [reply]
@javiercn, what CORS headers they need to send back?

**Reply by javiercn** [reply]
The CORS docs on MDN do a much better job than me at explaining this in depth https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
